### PR TITLE
PP-6262 Create Payouts Dao/Entity

### PR DESCRIPTION
**Pull request by rorymalcolm**
- Creates the PayoutEntity, PayoutDao and PayoutFixture to allow for the
representation of the `payouts` database table in Java objects.

- Adds tests to ensure the DAO methods we have chosed to start with work
as expected.

### CODE REVIEW

File: src/main/java/uk/gov/pay/ledger/payout/dao/PayoutDao.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            ":id, " +
+            ":gateway_payout_id, " +
+            ":amount, " +
+            ":created_date, " +
+            ":paid_out_date, " +
+            ":statement_descriptor, " +
+            ":status, " +
+            ":type" +
+            ") " +
+            "ON CONFLICT (id) DO UPDATE SET " +
```
[the remark this answers is not in the document]
Comment (rorymalcolm) [reply src/main/java/uk/gov/pay/ledger/payout/dao/PayoutDao.java]: Yeah agreed, sorry, it was late, I got confused!

File: src/main/java/uk/gov/pay/ledger/payout/dao/PayoutDao.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Optional;
+
+public class PayoutDao {
+
+    private final String SELECT_PAYOUT_BY_GATEWAY_PAYOUT_ID = "SELECT * FROM payouts " +
+            "WHERE gateway_payout_id = :gatewayPayoutId";
+
+    private final String UPSERT_PAYOUT = "INSERT INTO payouts " +
+            "(" +
+            "id,\n" +
```
[the remark this answers is not in the document]
Comment (rorymalcolm) [reply src/main/java/uk/gov/pay/ledger/payout/dao/PayoutDao.java]: They are used in the other queries so I copied the convention

File: src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    private final String statementDescriptor;
+    private final String status;
+    private final String type;
+    @JsonIgnore
+    private final Long version;
+
+    public PayoutEntity(Long id, String gatewayPayoutId, Long amount, ZonedDateTime createdDate,
+                        ZonedDateTime paidOutDate, String statementDescriptor, String status,
+                        String type, Long version) {
+        this.id = id;
```
[the remark this answers is not in the document]
Comment (rorymalcolm) [reply src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java]: even if we might want to get them at a later date?

File: src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    private final Long id;
+    private final String gatewayPayoutId;
+    private final Long amount;
+    private final ZonedDateTime createdDate;
+    private final ZonedDateTime paidOutDate;
+    private final String statementDescriptor;
+    private final String status;
+    private final String type;
+    @JsonIgnore
+    private final Long version;
```
[the remark this answers is not in the document]
Comment (rorymalcolm) [reply src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java]: should I create a migration to remove the field?

[the remark this answers is not in the document]
Comment (kbottla) [reply src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java]: yeah, a separate PR is fine though

File: src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    private final String gatewayPayoutId;
+    private final Long amount;
+    private final ZonedDateTime createdDate;
+    private final ZonedDateTime paidOutDate;
+    private final String statementDescriptor;
+    private final String status;
+    private final String type;
+    private final Long version;
+
+    public PayoutEntity(Long id, String gatewayPayoutId, Long amount,
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java]: Sorry missed out first time. We could actually pass `PayoutEntityBuilder` so constructor parameters are manageable and is easy to add more values if needed. Usual recommended by lint tools are a max of 7.

File: src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kbottla) [reply src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java]: 👍  fine if we are using it to map from DB

File: src/test/java/uk/gov/pay/ledger/payout/dao/PayoutDaoIT.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        assertThat(payout.getGatewayPayoutId(), is("po_test2"));
+        assertThat(payout.getStatementDescriptor(), is("a statement descriptor"));
+        assertThat(payout.getStatus(), is("PAYOUT"));
+        assertThat(payout.getType(), is("Bank Account"));
+        assertThat(payout.getCreatedDate(), is(notNullValue()));
+        assertThat(payout.getPaidOutDate(), is(paidOutDate));
+    }
+
+    @Test
+    public void shouldUpsertExistingTrasnaction() {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/ledger/payout/dao/PayoutDaoIT.java]:
```suggestion
    public void shouldUpsertExistingTransaction() {
```

File: src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java (hunk 5 above)
Comment (rorymalcolm) [reply src/main/java/uk/gov/pay/ledger/payout/entity/PayoutEntity.java]: will do!

**Pull request merged**